### PR TITLE
[JIT Kernel] Migrate per_token_quant_fp8 to JIT

### PR DESCRIPTION
**Pull request by xingsy97**
## Motivation

Migrate `per_token_quant_fp8` kernel to JIT compilation (#17865).

## Modifications

Under `python/sglang/jit_kernel/`:
- `csrc/gemm/per_token_quant_fp8.cuh` — CUDA kernel (ported from `sgl-kernel/csrc/gemm/per_token_quant_fp8.cu`)
- `per_token_quant_fp8.py` — Python wrapper
- `tests/test_per_token_quant_fp8.py` — Unit tests
- `benchmark/bench_per_token_quant_fp8.py` — Benchmark (JIT vs AOT vs PyTorch)

And
- `python/sglang/srt/layers/quantization/fp8_kernel.py` — Switch import to JIT

## Accuracy Tests

Pass all tests defined in `python/sglang/jit_kernel/tests/test_per_token_quant_fp8.py`

## Benchmarking and Profiling

```
GPU: NVIDIA GeForce RTX 5060 Laptop GPU
Driver: 580.126.20 | CUDA: 12.8 | PyTorch: 2.9.1+cu128
Script: python/sglang/jit_kernel/benchmark/bench_per_token_quant_fp8.py

per-token-quant-fp8-performance (unit: us):
   M × K (dtype)       JIT    AOT    PyTorch  AOT->JIT
   1 × 512, fp16       4.6    5.1    37.3     -9.8%
   1 × 4096, bf16      4.7    5.2    51.2     -9.6%
   16 × 2048, fp16     4.8    5.3    54.7     -9.4%
   64 × 4096, fp32     7.1    7.0    120.8    +1.4%
   128 × 7168, bf16    8.6    8.2    224.6    +4.9%
   256 × 7168, fp16    12.3   12.1   410.2    +1.7%
```

JIT and AOT performance are essentially the same, no regression observed. Both are significantly faster than PyTorch.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

### CODE REVIEW

File: python/sglang/jit_kernel/benchmark/bench_per_token_quant_fp8.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+        def fn():
+            absmax = input_tensor.float().abs().max(dim=-1, keepdim=True).values
+            scale = absmax / FP8_E4M3_MAX
+            scale_inv = torch.where(
+                scale == 0, torch.zeros_like(scale), 1.0 / scale
+            )
+            _ = (input_tensor.float() * scale_inv).clamp(
+                -FP8_E4M3_MAX, FP8_E4M3_MAX
+            )
```
[the remark this answers is not in the document]
Comment (xingsy97) [reply python/sglang/jit_kernel/benchmark/bench_per_token_quant_fp8.py]: fixed, now it writes to output_q and output_s like the kernels do.

File: python/sglang/jit_kernel/csrc/gemm/per_token_quant_fp8.cuh [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    if constexpr (kVecSize == 16) {
+      *reinterpret_cast<uint4*>(token_output + i * kVecSize) = *reinterpret_cast<uint4*>(out_arr);
+    } else {
+#pragma unroll
+      for (int k = 0; k < kVecSize; ++k) {
+        token_output[i * kVecSize + k] = out_arr[k];
+      }
+    }
```
[the remark this answers is not in the document]
Comment (xingsy97) [reply python/sglang/jit_kernel/csrc/gemm/per_token_quant_fp8.cuh]: added vectorized stores for kVecSize=8 (uint2) and kVecSize=4 (uint) too.

File: python/sglang/jit_kernel/csrc/gemm/per_token_quant_fp8.cuh [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+    fp8_e4m3_t out_arr[kVecSize];
+#pragma unroll
+    for (int j = 0; j < kVecSize; ++j) {
+      float val = static_cast<float>(in_vec[j]) * scale_inv;
+      val = math::max(math::min(val, math::FP8_E4M3_MAX), -math::FP8_E4M3_MAX);
+      out_arr[j] = static_cast<fp8_e4m3_t>(val);
+    }
+
+    if constexpr (kVecSize == 16) {
```
Comment (DarkSharpness) [inline python/sglang/jit_kernel/csrc/gemm/per_token_quant_fp8.cuh]: Why use reinterpret_cast here? Why not `AlignedVector`?

File: python/sglang/jit_kernel/csrc/gemm/per_token_quant_fp8.cuh [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+  const auto num_tokens = static_cast<int64_t>(M.unwrap());
+  const auto hidden_dim = static_cast<int64_t>(D.unwrap());
+
+  RuntimeCheck(hidden_dim % 4 == 0, "per_token_quant_fp8: hidden_dim must be divisible by 4, got ", hidden_dim);
+
+  constexpr int kBlockSize = 256;
+  constexpr int kMaxVecSize = 32 / sizeof(DType);
+
+  if (hidden_dim % kMaxVecSize == 0) {
```
Comment (DarkSharpness) [inline python/sglang/jit_kernel/csrc/gemm/per_token_quant_fp8.cuh:238]: Can we use some utility function to avoid passing the arguments same many times? The only difference here is the kernel.

e.g. use a local lambda functon which returns a pointer to the real kernel

Comment (xingsy97) [reply python/sglang/jit_kernel/csrc/gemm/per_token_quant_fp8.cuh:238]: Done. Extracted common logic into a lambda.

File: python/sglang/jit_kernel/csrc/gemm/per_token_quant_fp8.cuh (hunk 3 above)
Comment (xingsy97) [reply python/sglang/jit_kernel/csrc/gemm/per_token_quant_fp8.cuh]: Done. changed to `AlignedVector`

File: python/sglang/jit_kernel/csrc/gemm/per_token_quant_fp8.cuh [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+template <typename DType, int kVecSize>
+__global__ void per_token_quant_fp8_kernel(const DType* __restrict__ input,
+                                           fp8_e4m3_t* __restrict__ output_q,
+                                           float* __restrict__ output_s,
+                                           const int64_t hidden_dim,
+                                           const int64_t num_tokens) {
+  using namespace device;
+
+  const int64_t token_idx = blockIdx.x;
+  if (token_idx >= num_tokens) return;
```
Comment (HydraQYH) [inline python/sglang/jit_kernel/csrc/gemm/per_token_quant_fp8.cuh]: Is this redundant? The size of the grid is `num_tokens`.After removing this line of code, parameter `const int64_t num_tokens` is also redundant.

File: python/sglang/jit_kernel/csrc/gemm/per_token_quant_fp8.cuh [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+      .with_device(device)
+      .verify(output_s);
+
+  const auto num_tokens = static_cast<int64_t>(M.unwrap());
+  const auto hidden_dim = static_cast<int64_t>(D.unwrap());
+
+  RuntimeCheck(hidden_dim % 4 == 0, "per_token_quant_fp8: hidden_dim must be divisible by 4, got ", hidden_dim);
+
+  constexpr int kBlockSize = 256;
+  constexpr int kMaxVecSize = 32 / sizeof(DType);
```
Comment (HydraQYH) [inline python/sglang/jit_kernel/csrc/gemm/per_token_quant_fp8.cuh]: Why 32? Does this assume a Blackwell architecture?

File: python/sglang/jit_kernel/csrc/gemm/per_token_quant_fp8.cuh (hunk 5 above)
Comment (xingsy97) [reply python/sglang/jit_kernel/csrc/gemm/per_token_quant_fp8.cuh]: yes it is. removed.

File: python/sglang/jit_kernel/csrc/gemm/per_token_quant_fp8.cuh (hunk 6 above)
Comment (xingsy97) [reply python/sglang/jit_kernel/csrc/gemm/per_token_quant_fp8.cuh]: 32 bytes here matches the AOT original version. In [per_token_quant_fp8.cu L18](https://github.com/sgl-project/sglang/blob/main/sgl-kernel/csrc/gemm/per_token_quant_fp8.cu#L18), the default template parameter is `kVecSize = 16`. For `float16` (2 bytes per element), that's `16 * 2 = 32` bytes per vector load.

The formula `32 / sizeof(DType)` generalizes this across dtypes (e.g., `float32` gives `32 / 4 = 8`, also matching the AOT `use_vec8` path).

From my understanding, 32-byte vector loads work correctly on all architectures. On pre-Blackwell GPUs the compiler splits them into two 16-byte load instructions. If #19794 gets merged first, I can update this to use `device::kMaxVecBytes` introduced by it directly.

Comment (DarkSharpness) [reply python/sglang/jit_kernel/csrc/gemm/per_token_quant_fp8.cuh]: Let's wait for #19794. You may profile the performance first, just ensure changing from 32 -> 16 should not bring no performance regression.

Comment (xingsy97) [reply python/sglang/jit_kernel/csrc/gemm/per_token_quant_fp8.cuh]: posted benchmark result in https://github.com/sgl-project/sglang/pull/19702#issuecomment-4009862593

Comment (HydraQYH) [reply python/sglang/jit_kernel/csrc/gemm/per_token_quant_fp8.cuh]: > 32 bytes here matches the AOT original version. In [per_token_quant_fp8.cu L18](https://github.com/sgl-project/sglang/blob/main/sgl-kernel/csrc/gemm/per_token_quant_fp8.cu#L18), the default template parameter is `kVecSize = 16`. For `float16` (2 bytes per element), that's `16 * 2 = 32` bytes per vector load.
> 
> The formula `32 / sizeof(DType)` generalizes this across dtypes (e.g., `float32` gives `32 / 4 = 8`, also matching the AOT `use_vec8` path).
> 
> From my understanding, 32-byte vector loads work correctly on all architectures. On pre-Blackwell GPUs the compiler splits them into two 16-byte load instructions. If #19794 gets merged first, I can update this to use `device::kMaxVecBytes` introduced by it directly.

@xingsy97 Yes. I think using `device::kMaxVecBytes` is better.

**Comment by hnyls2002** [comment]
Thanks @xingsy97! The JIT tree was relocated from `python/sglang/jit_kernel/` into `sglang.kernels` by #31666, so every path in this diff (including the consumer `srt/layers/quantization/fp8_kernel.py`, now `kernels/ops/quantization/fp8_kernel.py`) no longer exists, and the `load_jit` API changed shape. To be clear the port itself is still wanted - `sgl_per_token_quant_fp8` is registered AOT-only at `python/sglang/kernels/ops/quantization/__init__.py:20-32`. Closing as stale against the new layout; a fresh PR adding `kernels/jit/csrc/gemm/per_token_quant_fp8.cuh` plus a `KernelBackend.JIT` registration would be very welcome. Please reopen if I've missed something.
